src/fundamental/chid-fundamental.c

```diff
@@ -0,0 +1,127 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+/*
+ * Based on Nikita Travkin's dtbloader implementation.
+ * Copyright (c) 2024 Nikita Travkin <nikita@trvn.ru>
+ *
+ * https://github.com/TravMurav/dtbloader/blob/main/src/chid.c
+ */
+
+/*
+ * Based on Linaro dtbloader implementation.
+ * Copyright (c) 2019, Linaro. All rights reserved.
+ *
+ * https://github.com/aarch64-laptops/edk2/blob/dtbloader-app/EmbeddedPkg/Application/ConfigTableLoader/CHID.c
+ */
+
+#if SD_BOOT
+#  include "efi-string.h"
+#else
+#  include <string.h>
+#  include <uchar.h>
+
+static inline size_t strlen16(const char16_t *s) {
+        if (!s)
+                return 0;
+
+        size_t len = 0;
+        while (*s) {
+                s++;
+                len++;
+        }
+
+        return len;
+}
```

Comment:
don't define this anew. instead, let's maybe just rename the existing char16_strlen() call in src/basic/utf8.h to strlen16() and use that here